pandas/tests/indexes/test_base.py

```diff
@@ -237,61 +223,63 @@ def test_constructor_int_dtype_float(self, dtype):
         result = Index([0., 1., 2., 3.], dtype=dtype)
         tm.assert_index_equal(result, expected)
 
-    def test_constructor_int_dtype_nan(self):
+    @pytest.mark.parametrize("dtype,klass_or_raises", [
+        ('int64', True), ('uint64', True), ('float', Float64Index)
+    ])
+    def test_constructor_int_dtype_nan(self, dtype, klass_or_raises):
         # see gh-15187
         data = [np.nan]
-        msg = "cannot convert"
-
-        with tm.assert_raises_regex(ValueError, msg):
-            Index(data, dtype='int64')
 
-        with tm.assert_raises_regex(ValueError, msg):
-            Index(data, dtype='uint64')
-
-        # This, however, should not break
-        # because NaN is float.
-        expected = Float64Index(data)
-        result = Index(data, dtype='float')
-        tm.assert_index_equal(result, expected)
+        if klass_or_raises is True:
+            msg = "cannot convert"
+            with tm.assert_raises_regex(ValueError, msg):
+                Index(data, dtype=dtype)
+        else:
+            expected = klass_or_raises(data)
+            result = Index(data, dtype=dtype)
+            tm.assert_index_equal(result, expected)
 
-    def test_index_ctor_infer_nan_nat(self):
+    @pytest.mark.parametrize("klass,dtype,na_val", [
+        (pd.Float64Index, np.float64, np.nan),
+        (pd.DatetimeIndex, 'datetime64[ns]', pd.NaT)
+    ])
+    def test_index_ctor_infer_nan_nat(self, klass, dtype, na_val):
         # GH 13467
-        exp = pd.Float64Index([np.nan, np.nan])
-        assert exp.dtype == np.float64
-        tm.assert_index_equal(Index([np.nan, np.nan]), exp)
-        tm.assert_index_equal(Index(np.array([np.nan, np.nan])), exp)
-
-        exp = pd.DatetimeIndex([pd.NaT, pd.NaT])
-        assert exp.dtype == 'datetime64[ns]'
-        tm.assert_index_equal(Index([pd.NaT, pd.NaT]), exp)
-        tm.assert_index_equal(Index(np.array([pd.NaT, pd.NaT])), exp)
-
+        na_list = [na_val, na_val]
+        exp = klass(na_list)
+        assert exp.dtype == dtype
+        tm.assert_index_equal(Index(na_list), exp)
+        tm.assert_index_equal(Index(np.array(na_list)), exp)
+
+    @pytest.mark.parametrize("data", [
+        [pd.NaT, np.nan], [np.nan, pd.NaT], [np.nan, np.datetime64('nat')],
```

Comment:
we should make this into a fixture (e.g. nulls_fixture) (if you want to add to conftest). note we should change globally (but can do that in another PR)